src/EventSubscriber/ApqSubscriber.php

```diff
@@ -44,12 +44,21 @@ public function onBeforeOperation(OperationEvent $event): void {
     $query = $event->getContext()->getOperation()->query;
     $queryHash = $event->getContext()->getOperation()->extensions['persistedQuery']['sha256Hash'] ?? '';
 
-    if (is_string($query) && is_string($queryHash) && $queryHash !== '') {
-      $computedQueryHash = hash('sha256', $query);
-      if ($queryHash !== $computedQueryHash) {
-        throw new Error('Provided sha does not match query');
+    if (is_string($queryHash) && $queryHash !== '') {
+      // Add cache context for dynamic page cache compatibility on all
+      // operations that have the hash set.
+      $event->getContext()->addCacheContexts(
+        ['url.query_args:variables', 'url.query_args:extensions']
+      );
+
+      // If we have a query and the hash matches then can cache it.
+      if (is_string($query)) {
+        $computedQueryHash = hash('sha256', $query);
+        if ($queryHash !== $computedQueryHash) {
+          throw new Error('Provided sha does not match query');
```

Comment:
I think Error is not correct here, should be UserError, right? We want to show this error to the graphql client that they sent something wrong.

Not the fault of this pull request, was already wrong before.

Reply:
Interestingly we are asserting in the tests that the error message is coming out in the response, so it appears to be treated already as user error for some reason ... so we can leave this as is and don't care for now.